### PR TITLE
Add single sensor mode to the sensor container service / process

### PR DESCRIPTION
**Pull request by Kami**
This pull request adds single sensor mode to the sensor container service (enabled via ``sensorcontainer.single_sensor_mode`` config option).

When this flag is set, sensor container process doesn't manage lifecycle of a sensor processes and respawn them if a sensor exits / crashes, but it exits itself and propagates sensor exit code to the parent sensor container process.

This comes handy in scenarios when a sensor life cycle and failover is handled by an external service such as Kubernetes / Mesos / monitoring + bash scripts / ...

Note: This flag needs to be used in combination with ``--sensor-ref`` CLI option since without it it doesn't make sense.

## TODO

- [x] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by armab** [comment]
Checked, that looks like what we need from the description :+1: 

Also sounds like sensor_partitioner is irrelevant in this mode?
How `sensorcontainer.sensor_node_name` works in this mode when specified in st2.conf?
Since in both cases sensor ref is already specified explicitly.

I will test/play more closely with the package produced from this build later next week.

File: st2reactor/st2reactor/sensor/config.py [hunk 1]
```diff
@@ -63,13 +64,28 @@ def _register_sensor_container_opts(ignore_errors=False):
...
 
+    # Other options
     st2cfg.do_register_opts(partition_opts, group='sensorcontainer', ignore_errors=ignore_errors)
 
-    sensor_test_opt = cfg.StrOpt(
-        'sensor-ref',
-        help='Only run sensor with the provided reference. Value is of the form pack.sensor-name.')
+    other_opts = [
+        cfg.BoolOpt(
+            'single_sensor_mode', default=False,
```
Comment (Kami) [inline st2reactor/st2reactor/sensor/config.py]: I'm also open to a better name for this option.

**Comment by Kami** [comment]
> Also sounds like sensor_partitioner is irrelevant in this mode?
Additionally, how sensorcontainer.sensor_node_name works (clashes, ignored?) in this mode when specified in st2.conf?

Correct, right now when ``--sensor-ref`` option is provided, partitioner settings are ignored and ``SingleSensorPartitioner`` is used.

I can modify the code a bit to at least log a message in this case since right now we just ignore partitioner setting in such scenarios without logging anything.

File: st2reactor/st2reactor/container/partitioner_lookup.py [hunk 2]
```diff
@@ -41,16 +41,19 @@
 
 def get_sensors_partitioner():
     if cfg.CONF.sensor_ref:
+        LOG.info('Running in single sensor mode, using a single sensor partitioner...')
```
Comment (armab) [inline st2reactor/st2reactor/container/partitioner_lookup.py]: :+1:

**Comment by Kami** [comment]
@armab Tests are now in place and this PR is ready to be reviewed although I'm still battling with getting integration tests to work on Travis under Python 3.

**Comment by armab** [comment]
```
$ kubectl logs githubwebhooksensor-c9cb6b84b-xdk97 -f
Adding password for user admin
/st2-docker/bin/entrypoint.sh: ignoring /st2-docker/entrypoint.d/*

2018-06-19 16:02:36,142 DEBUG [-] Using config files: /etc/st2/st2.conf
2018-06-19 16:02:36,143 DEBUG [-] Using logging config: /etc/st2/logging.sensorcontainer.conf
2018-06-19 16:02:36,144 INFO [-] Connecting to database "st2" @ "mongo:27017" as user "None".
2018-06-19 16:02:36,316 INFO [-] Successfully connected to database "st2" @ "mongo:27017" as user "None".
2018-06-19 16:02:36,605 INFO [-] Running in single sensor mode, using a single sensor partitioner...
2018-06-19 16:02:36,607 INFO [-] Setting up container to run 1 sensors.
2018-06-19 16:02:36,607 INFO [-] 	Sensors list - [u'githubwebhook.GitHubWebhookSensor'].
2018-06-19 16:02:36,608 INFO [-] (PID:20) SensorContainer started.
2018-06-19 16:02:36,608 INFO [-] Running sensor githubwebhook.GitHubWebhookSensor
2018-06-19 16:02:36,620 AUDIT [-] Access granted to "sensors_container" with the token set to expire at "2018-06-20T16:02:36.617606Z". (username='sensors_container',token_expiration='2018-06-20T16:02:36.617606Z')
2018-06-19 16:02:36,672 INFO [-] Connected to amqp://admin:**@rabbitmq:5672//
2018-06-19 16:02:36,734 INFO [-] Sensor githubwebhook.GitHubWebhookSensor started
2018-06-19 16:02:37,568 INFO [-] Found config for sensor "GitHubWebhookSensor"
2018-06-19 16:02:37,571 INFO [-] Watcher started
2018-06-19 16:02:37,571 INFO [-] Running sensor initialization code
2018-06-19 16:02:37,573 INFO [-] Running sensor in passive mode
 * Serving Flask app "webhook_sensor" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: on
2018-06-19 16:02:37,584 INFO [-]  * Running on http://0.0.0.0:8642/ (Press CTRL+C to quit)
2018-06-19 16:02:37,585 INFO [-]  * Restarting with stat
2018-06-19 16:02:37,655 INFO [-] Connected to amqp://admin:**@rabbitmq:5672//
2018-06-19 16:02:38,489 INFO [-] Found config for sensor "GitHubWebhookSensor"
2018-06-19 16:02:38,490 INFO [-] Watcher started
2018-06-19 16:02:38,490 INFO [-] Running sensor initialization code
2018-06-19 16:02:38,492 INFO [-] Running sensor in passive mode
2018-06-19 16:02:38,497 WARNING [-]  * Debugger is active!
2018-06-19 16:02:38,616 INFO [-]  * Debugger PIN: 190-833-188
2018-06-19 16:02:38,691 INFO [-] Connected to amqp://admin:**@rabbitmq:5672//
2018-06-19 18:07:39,841 INFO [-] Stopping trigger watcher
2018-06-19 18:07:39,848 INFO [-] Invoking cleanup on sensor
2018-06-19 18:07:40,133 INFO [-] Stopping trigger watcher
2018-06-19 18:07:40,135 INFO [-] Invoking cleanup on sensor
2018-06-19 18:07:41,081 INFO [-] Process for sensor githubwebhook.GitHubWebhookSensor has exited with code 0
2018-06-19 18:07:41,086 INFO [-] Not respawning a sensor since running in single sensor mode (sensor_id=u'githubwebhook.GitHubWebhookSensor',sensor={'class_name': u'GitHubWebhookSensor', 'poll_interval': None, 'pack': u'githubwebhook', 'ref': u'githubwebhook.GitHubWebhookSensor', 'file_path': u'/opt/stackstorm/packs/githubwebhook/sensors/webhook_sensor.py', 'trigger_types': [u'githubwebhook.github_event']})
2018-06-19 18:07:42,086 ERROR [-] Process container stopped.
2018-06-19 18:07:42,086 ERROR [-] Process container quit with exit_code 0.
2018-06-19 18:07:42,086 ERROR [-] (PID:20) SensorContainer stopped.

```

:+1:

**Comment by armab** [comment]
Just a future note for myself, when I'll need to productionize that in Docker & K8s: 
```sh
exec /opt/stackstorm/st2/bin/st2sensorcontainer \
  --config-file /etc/st2/st2.conf \
  --single-sensor-mode \
  --sensor-ref=githubwebhook.GitHubWebhookSensor
```

**Comment by lakshmi-kannan** [comment]
I don't think we should merge this for 2.8.0 but instead bundle all HA changes for 2.9.0.

**Comment by Kami** [comment]
@lakshmi-kannan It doesn't negatively affect things (or change the behavior) in any way I see no reason to not merge it.

**Comment by Kami** [comment]
@armab Yeah, you can also set single sensor mode via config option, but I personally prefer the command line flag since it's more explicit and visible compared to the config file option.

**Comment by lakshmi-kannan** [comment]
@Kami If you feel strongly and you are sure it won't break anything with sensor behavior today, I am +1 to shipping this. I looked at the code and I don't expect it to break existing behavior. Please validate this and you have my +1.

However, splitting timers into own process though isn't functionally different, I will wait until after 2.8 to ship it.

**Comment by Kami** [comment]
@lakshmi-kannan Yeah, I agree about the timer change - it's more involved since it includes a new service and packaging + installation changes.

**Pull request merged**